python/cudf_polars/cudf_polars/experimental/sort.py

```diff
@@ -4,19 +4,350 @@
+
+        by = options["by"]
+
+        splits = find_sort_splits(
+            df.select(by).table,
+            sort_boundaries.table,
+            partition_id,
+            partition_count,
+            options["order"],
+            options["null_order"],
+        )
+        packed_inputs = split_and_pack(
+            df.table,
+            splits=splits,
+            num_partitions=partition_count,
+            stream=DEFAULT_STREAM,
+            device_mr=rmm.mr.get_current_device_resource(),
+        )
+        shuffler.insert_chunks(packed_inputs)
+
+    @staticmethod
+    def extract_partition(
+        partition_id: int,
+        shuffler: Any,
+        options: SortedShuffleOptions,
+    ) -> DataFrame:
+        """Extract a finished partition from the RMP shuffler."""
+        from rapidsmpf.shuffler import unpack_and_concat
+
+        shuffler.wait_on(partition_id)
+        column_names = options["column_names"]
+
+        # TODO: When sorting, this step should finalize with a merge (unless we
+        # require stability, as cudf merge is not stable).
+        return DataFrame.from_table(
+            unpack_and_concat(
+                shuffler.extract(partition_id),
+                stream=DEFAULT_STREAM,
+                device_mr=rmm.mr.get_current_device_resource(),
+            ),
+            column_names,
+        )
+
+
+def _sort_partition_dataframe(
+    df: DataFrame,
+    partition_id: int,  # Not currently used
+    partition_count: int,
+    options: MutableMapping[str, Any],
+    sort_boundaries: DataFrame,
+) -> MutableMapping[int, DataFrame]:
+    """
+    Partition a sorted DataFrame for shuffling.
+
+    Parameters
+    ----------
+    df
+        The DataFrame to partition.
+    partition_id
+        The partition id of the current partition.
+    partition_count
+        The total number of partitions.
+    options
+        The sort options ``(by, order, null_order)``.
+    sort_boundaries
+        The global sort boundary candidates used to decide where to split.
+    """
+    if df.num_rows == 0:
+        # Fast path for empty DataFrame
+        return {i: df for i in range(partition_count)}
+
+    splits = find_sort_splits(
+        df.select(options["by"]).table,
+        sort_boundaries.table,
+        partition_id,
+        partition_count,
+        options["order"],
+        options["null_order"],
+    )
+
+    # Split and return the partitioned result
+    return {
+        i: DataFrame.from_table(
+            split,
+            df.column_names,
+        )
+        for i, split in enumerate(plc.copying.split(df.table, splits))
+    }
+
+
+class ShuffleSorted(IR):
+    """
+    Shuffle sorted multi-partition data.
+
+    Shuffling is performed by extracting sort boundary candidates from all partitions,
+    sharing them all-to-all and then exchanging data accordingly.
+    The sorting information is required to be passed in identically
+    to the initial sort.
```

Comment:
Can we somehow establish as clearly as possible that the child of this IR node must have locally sorted partitions? I think you have stated this both implicitly and explicitly in other places.  I suspect it would be good information to have here as well.